### PR TITLE
fix: (datagrid) filtering not working well with Ivy off

### PR DESCRIPTION
**Pull request by Jinnie**
issue #5085

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] clarity.design website / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

When Ivy is turned off, the displayedRows list accumulates old ghost items, which mess the proper child views visualization. Thus, after been filtered out, rows are not shown back when the filter is removed.

Issue Number: #5085 

## What is the new behavior?

Works fine both with and without Ivy

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Please note, that because of the complicated setup we can't make a test for the no-Ivy scenario. We can't have a test run with and without Ivy unless we run a new task with new test setup.

### CODE REVIEW

**Review by bbogdanov — DISMISSED** [review]

**Review by sis0k0 — APPROVED** [review]

**Review by gnomeontherun — COMMENTED** [review]
We've had a regression from fixes here before. I want to be 100% certain before we merge that this is correct. How have we tested this?

**Comment by Jinnie** [comment]
As I said, automated tests are not possible, so:
1. Checked with enabled Ivy that it does not break any current tests. I didn't test with Ivy off, as many of our other tests are failing with Ivy off.
2. Manually tested the following scenarios (both Ivy on and off):
    - no interference with the trackby functionalities - it keeps the proper items selected.
    - made sure the recently fixed issue of keeping detail pane open on refresh is okay.
3. Made sure we're not skipping the view destruction, which was previously leading to memory leaks.
4. Debugged the previously offending data list, which was accumulating dead records, thus leading to possible memory leaks.

Any other testing ideas or manually performed testing by the reviewers are welcome. I won't merge this for a couple of day more to allow for more reviews.

**Comment by Jinnie** [comment]
After today's tests I made a more precise selection of what should be removed - only the views marked as "destroyed". Without it - it was possible to get "pseudo-randomly" rearranged rows after update.

**Comment by gnomeontherun** [comment]
Thanks, I wanted to be as certain as we can. It helps to put more of this detail into the issue PR description as well in the future.

**Comment by Jinnie** [comment]
Thanks form me too - it was good to have one more round - we found an issue. I'm postponing the unblocking with 1 more day.

File: packages/angular/projects/clr-angular/src/data/datagrid/datagrid.ts
Comment (Shijir) [inline packages/angular/projects/clr-angular/src/data/datagrid/datagrid.ts:233]: I think this was due to an Angular bug—something along the lines of this: https://github.com/angular/angular/issues/38648. 

When `ViewRef` destroys itself independently without going through `ViewContainerRef` as we do in our `WrappedRow`'s `OnDestroy` cycle, the `ViewContainerRef` still holds the reference to the `ViewRef`. And that results in out-of-sync views and `ViewContainerRef`.

**Review by Shijir — DISMISSED** [review]

File: packages/angular/projects/clr-angular/src/data/datagrid/datagrid.ts
Comment (Jinnie) [reply packages/angular/projects/clr-angular/src/data/datagrid/datagrid.ts:233]: I think I started with the `ViewContainerRef.remove()` at first, but it was having some side effects. Then I saw that in other part of this same TS file we're using detach() and it worked better. But I'm not sure if I was doing the `.destroyed` check back then, so I'll give it a try, out of curiosity at least. 
Thanks for the detailed review, Shijir!

Comment (Jinnie) [reply packages/angular/projects/clr-angular/src/data/datagrid/datagrid.ts]: You're right, though it probably gets reclaimed after being detached, using remove gives us finer level of control and better protection against memory leaks. Updated all 3 PRs to use remove()

**Comment by hippee-lee** [comment]
@Jinnie - fyi I updated the base branch to reflect the new default for the repo.

**Review by Shijir — APPROVED** [review]
I will approve it again. I am only a bit apprehensive we couldn't test it automatically. But I am sure @Jinnie tested well with both Ivy and Ivy off.

**Review by bbogdanov — APPROVED** [review]
What do you think about creating an issue or a discussion to revisit this code at some point in order to covert it with tests?

**Pull request merged**
